src/node.rs

```diff
@@ -114,16 +135,250 @@ impl InMemoryNodeInner {
+
+        if pubdata_for_factory_deps > MAX_PUBDATA_PER_BLOCK {
+            return Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                "exceeds limit for published pubdata".into(),
+                Default::default(),
+            )));
+        }
+
+        let gas_for_bytecodes_pubdata: u32 =
+            pubdata_for_factory_deps * (gas_per_pubdata_byte as u32);
+
+        // We are using binary search to find the minimal values of gas_limit under which the transaction succeeds
+        let mut lower_bound = 0;
+        let mut upper_bound = MAX_L2_TX_GAS_LIMIT as u32;
+
+        while lower_bound + ESTIMATE_GAS_ACCEPTABLE_OVERESTIMATION < upper_bound {
+            let mid = (lower_bound + upper_bound) / 2;
+            let try_gas_limit = gas_for_bytecodes_pubdata + mid;
+
+            let estimate_gas_result = self.estimate_gas_step(
+                l2_tx.clone(),
+                gas_per_pubdata_byte,
+                try_gas_limit,
+                l1_gas_price,
+                base_fee,
+            );
+            if estimate_gas_result.is_err() {
+                lower_bound = mid + 1;
+            } else {
+                upper_bound = mid;
+            }
+        }
+
+        let tx_body_gas_limit = cmp::min(
+            MAX_L2_TX_GAS_LIMIT as u32,
+            (upper_bound as f32 * ESTIMATE_GAS_SCALE_FACTOR) as u32,
+        );
+        let suggested_gas_limit = tx_body_gas_limit + gas_for_bytecodes_pubdata;
+
+        let estimate_gas_result = self.estimate_gas_step(
+            l2_tx.clone(),
+            gas_per_pubdata_byte,
+            suggested_gas_limit,
+            l1_gas_price,
+            base_fee,
+        );
+
+        match estimate_gas_result {
+            Err(_) => Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                "Transaction is unexecutable".into(),
+                Default::default(),
+            ))),
+            Ok(_) => {
+                let overhead: u32 = derive_gas_estimation_overhead(
+                    suggested_gas_limit,
+                    gas_per_pubdata_byte as u32,
+                    tx.encoding_len(),
+                );
+
+                let full_gas_limit =
+                    match tx_body_gas_limit.overflowing_add(gas_for_bytecodes_pubdata + overhead) {
+                        (value, false) => value,
+                        (_, true) => {
+                            return Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                                "exceeds block gas limit".into(),
+                                Default::default(),
+                            )))
+                        }
+                    };
+
+                let fee = Fee {
+                    max_fee_per_gas: base_fee.into(),
+                    max_priority_fee_per_gas: 0u32.into(),
+                    gas_limit: full_gas_limit.into(),
+                    gas_per_pubdata_limit: gas_per_pubdata_byte.into(),
+                };
```

Comment:
and also print into the logs more detailed info about this

Reply:
This is the success response, I don't see value to print this in the logs at the moment.

We have another effort where we'll want to add debug/trace logs throughout. The first ones will be logs for all `eth_` requests and responses, so this can be included there (and this is for happy-path, where as errors we can totally add `println!()` for now)